### ISSUE TITLE
00-eda-airbnb-ny.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/abgonzalez-introduction-machine-learning#1]
Hola @abgonzalez , tu EDA es **excelente en general**, te hablo un poco del **data leakage** que debemos corregir. Tu análisis exploratorio está muy bien hecho, pero justo antes del modelado necesitas ajustar el orden de operaciones.

---

## **Cuidado: Data Leakage en Scaling**

Veo que en tu sección 8.3, haces scaling **antes** de hacer el split train/test:

**ACTUAL:**
```python
# Sección 8.3
scaler = MinMaxScaler()
scaled_features = scaler.fit_transform(df_no_outliers[num_variables_to_scale])  # ← Fit sobre TODOS los datos

df_final = df_scaled.copy()
# ... luego haces split en sección 9
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
```

**PROBLEMA:**
El scaler aprende los valores mín/máx de **TODO el dataset**, incluyendo el test set. Esto podría causar **data leakage**: estadísticas del test set influyen en la transformación del train set, haciendo que el modelo parezca mejor de lo que realmente es en ocasiones.

---

## **CORRECCIÓN - El Orden Correcto:**

Necesitas **reorganizar la sección 9** para que sea así:

```python
from sklearn.feature_selection import f_regression, SelectKBest
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler

# 1 PREPARAR FEATURES Y TARGET (sin escalar todavía)
# Usamos df_no_outliers directamente, sin escalado previo
num_variables_to_scale = ["latitude", "longitude", "price", "minimum_nights", 
                          "number_of_reviews", "calculated_host_listings_count", 
                          "availability_365"]

# Variables a usar (numéricas sin escalar + categóricas factorizadas)
features_cols = num_variables_to_scale + ["room_type_factor", "neighbourhood_group_factor"]

X = df_no_outliers[features_cols]
y = df_no_outliers["price"]

# 2 SPLIT PRIMERO (antes de scaling)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 3 SCALING - Fit SOLO en train
scaler = MinMaxScaler()
scaler.fit(X_train[num_variables_to_scale])  # ← Fit SOLO en train

# 4 Transform en train y test por separado
X_train_scaled = X_train.copy()
X_test_scaled = X_test.copy()

X_train_scaled[num_variables_to_scale] = scaler.transform(X_train[num_variables_to_scale])
X_test_scaled[num_variables_to_scale] = scaler.transform(X_test[num_variables_to_scale])

# 5 FEATURE SELECTION
selection_model = SelectKBest(f_regression, k=4)
selection_model.fit(X_train_scaled, y_train)
ix = selection_model.get_support()

# Crear DataFrames con las 4 mejores características
X_train_sel = X_train_scaled[X_train_scaled.columns[ix]]
X_test_sel = X_test_scaled[X_test_scaled.columns[ix]]

print(f"Features originales: {X_train_scaled.shape[1]}")
print(f"Features seleccionadas: {X_train_sel.shape[1]}")
print(f"\n Características seleccionadas:")
for col in X_train_sel.columns:
    print(f"   • {col}")

# GUARDAR DATASETS
os.makedirs("../data/processed", exist_ok=True)

X_train_sel.to_csv("../data/processed/X_train_airbnb.csv", index=False)
X_test_sel.to_csv("../data/processed/X_test_airbnb.csv", index=False)
y_train.to_csv("../data/processed/y_train_airbnb.csv", index=False)
y_test.to_csv("../data/processed/y_test_airbnb.csv", index=False)

print("\n Datasets guardados correctamente en ../data/processed/")
```

---

## **Lo que está bien en tu notebook:**

- **Análisis exploratorio impecable**: visualizaciones claras, interpretaciones correctas
- **Eliminación correcta de outliers** con método IQR bajo decisión
- **Factorización de categóricas** bien hecha
- **Feature selection con SelectKBest** en el orden correcto
- **Documentación excelente** con explicaciones claras

---

Genial el doble enfoque para comparar, vamos a tomar el primer enfoque ante todo que es en el que nosotros tomamos decisiones conociendo los datos paso a paso pero me parece genial que utilices ambos.

---

**Entonces:** El 95% de tu trabajo es excelente. Solo necesitas reorganizar esas 3 celdas del final (secciones 8.3-9) para que el scaling sea después del split train/test. Hay diversas vertientes de este procedimiento por lo que he aprendido de distintas personas ese es el que menos riesgo tiene respecto a la realidad de nuestros datos.

**Comment by abgonzalez** [comment]
Hola @mezcolantriz , 

Muchas gracias por tu trabajo. Revisandolo otra vez, he visto alguna cosa mas a mejorar tambien. 

Por favor, ¿Podiras echarle un vistazo?

Saludos y gracias.
